### PR TITLE
Promo coupons

### PR DESCRIPTION
**Pull request by chaotixkilla**

### CODE REVIEW

**Comment by chaotixkilla** [comment]
[Promo codes PT5: Refactor code related to bills, transaction value and rates](https://trello.com/c/p85Dgtwf/359-promo-codes-pt5-refactor-code-related-to-bills-transaction-value-and-rates)

**Comment by chaotixkilla** [comment]
[Promo codes PT4: Add the coupon flow to the machine logic](https://trello.com/c/p7UW5YBv/315-promo-codes-pt4-add-the-coupon-flow-to-the-machine-logic)

**Comment by chaotixkilla** [comment]
[Promo codes PT2: Create the machine screens.](https://trello.com/c/Y1zOCSns/313-promo-codes-pt2-create-the-machine-screens)

**Comment by chaotixkilla** [comment]
[Promo codes PT1: Create the coupon admin screen.](https://trello.com/c/As4wmMfA/301-promo-codes-pt1-create-the-coupon-admin-screen)

File: ui/css/main.scss [hunk 1]
```diff
@@ -586,7 +586,7 @@ footer {
   margin-top: auto;
 }
 
-.choose_fiat_state ,.insert_more_bills_state {
+.choose_fiat_state /*, .insert_more_bills_state */ {
```
Comment (csrapr) [inline ui/css/main.scss:589]: delete annotation

File: ui/js/app.js [hunk 2]
```diff
@@ -1662,4 +1700,32 @@ function shouldEnableTouch() {
...
+    $('#choose-fiat-coupon-added').show();
+  } else {
+    $('#insert-first-bill-promo-button').show();
+    $('#choose-fiat-promo-button').show();
+    $('#insert-first-bill-coupon-added').hide();
+    $('#choose-fiat-coupon-added').hide();
+  }
+}
+
 //# sourceMappingURL=app.js.map
```
Comment (csrapr) [inline ui/js/app.js:1731]: remove annotation and add newline at the end of the file

File: ui/src/app.js [hunk 3]
```diff
@@ -135,6 +138,8 @@ function processData (data) {
   if (data.operatorInfo) setOperatorInfo(data.operatorInfo)
   if (data.hardLimit) setHardLimit(data.hardLimit)
   if (data.cryptomatModel) setCryptomatModel(data.cryptomatModel)
+  if (data.areThereAvailableCoupons !== undefined) setAvailableCoupons(data.areThereAvailableCoupons)
```
Comment (csrapr) [inline ui/src/app.js:141]: can this be like this instead? `if (data.areThereAvailableCoupons) setAvailableCoupons(data.areThereAvailableCoupons)`

**Review by csrapr — CHANGES_REQUESTED** [review]

File: ui/js/app.js (hunk 2 above)
Comment (chaotixkilla) [reply ui/js/app.js:1731]: This is an auto-generated file, so I won't touch it

File: ui/src/app.js (hunk 3 above)
Comment (chaotixkilla) [reply ui/src/app.js:141]: it is a boolean, `if (data.areThereAvailableCoupons)` won't distinguish between undefined and false.

File: ui/js/app.js [hunk 4]
```diff
@@ -302,10 +313,10 @@ function chooseCoin(coins, twoWayMode) {
   }
 
   isTwoWay = twoWayMode;
-  setChooseCoinColors();
+  setChooseCoinColors
   // setupAnimation(twoWayMode, aspectRatio800)
 
-  var defaultCoin = coins[0];
+  ();var defaultCoin = coins[0];
```
Comment (csrapr) [inline ui/js/app.js:319]: is this on purpose?

File: ui/js/app.js [hunk 5]
```diff
@@ -440,17 +451,19 @@ $(document).ready(function () {
   window.addEventListener('resize', function () {
     calculateAspectRatio();
     setChooseCoinColors();
-  });
+  }
 
   // Matt's anti-drag hack
-  window.onclick = window.oncontextmenu = window.onmousedown = window.onmousemove = window.onmouseup = function () {
+  );window.onclick = window.oncontextmenu = window.onmousedown = window.onmousemove = window.onmouseup = function () {
```
Comment (csrapr) [inline ui/js/app.js:457]: also this

File: ui/js/app.js [hunk 6]
```diff
@@ -545,9 +558,29 @@ $(document).ready(function () {
...
+  setupButton('submit-promo-code', 'submitPromoCode', {
+    input: $('.promo-code-input').data('content')
+  });
+  setupButton('promo-code-try-again', 'insertPromoCode');
+  setupButton('promo-code-continue', 'cancelPromoCode');
+
+  setupButton('initialize', 'initialize'
   // setupButton('test-mode', 'testMode')
-  setupButton('pairing-scan', 'pairingScan');
+  );setupButton('pairing-scan', 'pairingScan');
```
Comment (csrapr) [inline ui/js/app.js:583]: out of place parentheses

File: ui/js/app.js [hunk 7]
```diff
@@ -730,10 +763,10 @@ function touchEvent(element, callback) {
   function handler(e) {
     var target = targetButton(e.target);
 
-    target.classList.add('active');
+    target.classList.add('active'
 
     // Wait for transition to finish
-    setTimeout(function () {
+    );setTimeout(function () {
```
Comment (csrapr) [inline ui/js/app.js:769]: same here

File: ui/js/app.js [hunk 8]
```diff
@@ -1046,10 +1080,10 @@ function setLocale(data) {
   var isHebrew = jsLocaleCode.indexOf('he-') === 0;
   isRTL = isArabic || isHebrew;
 
-  setChooseCoinColors();
+  setChooseCoinColors
   // setupAnimation(isTwoWay, aspectRatio800)
 
-  if (isRTL) {
+  ();if (isRTL) {
```
Comment (csrapr) [inline ui/js/app.js:1086]: here as well

File: ui/js/app.js [hunk 9]
```diff
@@ -1261,9 +1297,9 @@ function setExchangeRate(_rates) {
 }
 
 function qrize(text, target, color, lightning) {
-  var image = document.getElementById('bolt-img');
+  var image = document.getElementById('bolt-img'
   // Hack for surf browser
-  var size = document.body.clientHeight * 0.36;
+  );var size = document.body.clientHeight * 0.36;
```
Comment (csrapr) [inline ui/js/app.js:1302]: here

File: ui/js/app.js [hunk 10]
```diff
@@ -1423,10 +1459,10 @@ function translatePage() {
     var el = $(this);
     var base = el.data('baseTranslation');
     el.attr('placeholder', locale.translate(base).fetch());
-  });
+  }
 
   // Adjust send coins button
-  var length = $('#send-coins span').text().length;
+  );var length = $('#send-coins span').text().length;
```
Comment (csrapr) [inline ui/js/app.js:1465]: here

File: ui/src/app.js [hunk 11]
```diff
@@ -1728,3 +1766,31 @@ function shouldEnableTouch () {
...
+    $('#choose-fiat-coupon-added').html(`✔ Coupon added (${currentDiscount}% discount)`)
+    $('#insert-first-bill-coupon-added').show()
+    $('#choose-fiat-coupon-added').show()
+  } else {
+    $('#insert-first-bill-promo-button').show()
+    $('#choose-fiat-promo-button').show()
+    $('#insert-first-bill-coupon-added').hide()
+    $('#choose-fiat-coupon-added').hide()
+  }
+}
```
Comment (csrapr) [inline ui/src/app.js:1796]: add newline at the end

**Review by csrapr — CHANGES_REQUESTED** [review]

File: ui/js/app.js (hunk 4 above)
Comment (chaotixkilla) [reply ui/js/app.js:319]: As we spoke the other day, it's an auto-generated file. Going to resolve all convos related to this file

**Review by csrapr — APPROVED** [review]

**Review by csrapr — APPROVED** [review]

File: lib/brain.js [hunk 12]
```diff
@@ -2762,9 +2787,92 @@ Brain.prototype.clearBill = function clearBill () {
...
+    ? 'acceptingFirstBill'
+    : 'acceptingBills'
+
+  const callback = returnState === 'acceptingFirstBill'
+    ? this._idle.bind(this)
+    : this._sendCoins.bind(this)
+
+  passTx
+    ? this._transitionState(returnState, { tx: this.tx })
+    : this._transitionState(returnState)
```
Comment (RafaelTaranto) [inline lib/brain.js:2860]: What happens if we pass the tx everytime?

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: lib/brain.js (hunk 12 above)
Comment (chaotixkilla) [reply lib/brain.js:2860]: hmm, I just tested this and I seems that on both cashin and cashout, passing the transaction every time does not affect anything. Might have been some overengineering on my part, back when promo codes could be inserted after the first bill. Removing this and related code.

**Review by RafaelTaranto — APPROVED** [review]

**Pull request merged**